### PR TITLE
Fdo part2

### LINKED ISSUE #68: continue modularization
We are coming to the point where we should do registrations and callouts, like the following:

```
if WANT_FDO:
   create['fdo'] = fdo_create_device
   delete['fdo'] = fdo_delete_device
...
```
and then in scim.py:

```
for a in create:
   create[a](...)
```

For models.py, we probably want to move things like fdo back to their own scim_fdo files, etc.

### PR DESCRIPTION
**Pull request by elear**
Corrects a JSON problem with a key (can't have line breaks).
Adds initial FDO call out.

This commit now also fixes #68.

### CODE REVIEW

**Comment by elear** [comment]
Fixes #29

File: gateway/app.py [hunk 1]
```diff
@@ -28,7 +28,10 @@
 from models import EndpointApp, OnboardingAppKey
 from util import make_hash
 from app_factory import create_app
-
+# pylint: disable-next=unused-import
+from scim_ethermab import EtherMABExtension
+# pylint: disable-next=unused-import
+from scim_fdo import FDOExtension
```
Comment (rohitmohan96) [inline gateway/app.py:34]: Why is this required?

File: gateway/models.py [hunk 2]
```diff
@@ -113,10 +111,9 @@ def serialize(self):
             }
         if self.ble_extension:
             response.update(self.ble_extension.serialize())
-        if self.ethermab_extension:
-            response.update(self.ethermab_extension.serialize())
-        if self.fdo_extension:
-            response.update(self.fdo_extension.serialize())
+        for ext in scim_ext_read:
+            ext(self,response)
```
Comment (rohitmohan96) [inline gateway/models.py:115]: Would prefer something like this instead of `ext`. 
```suggestion
        for read_fn in scim_ext_read:
            read_fn(self, response)
```
Same comment for the other places where you have `ext`.

File: gateway/models.py [hunk 3]
```diff
@@ -113,10 +111,9 @@ def serialize(self):
             }
         if self.ble_extension:
             response.update(self.ble_extension.serialize())
```
Comment (rohitmohan96) [inline gateway/models.py:113]: Why is the ble extension not in scim_ext_read and handled separately like this?

File: gateway/scim.py [hunk 4]
```diff
@@ -259,14 +249,17 @@ def delete_device(entry_id):
     entry = session.get(Device,entry_id)
     if not entry:
         return blow_an_error("Device not found",404)
-    if entry.ethermab_extension:
-        ethermab_delete_device(entry.ethermab_extension)
 
-    for cls in [ BleExtension, EtherMABExtension, FDOExtension ]:
+    for cls in [ BleExtension ]:
```
Comment (rohitmohan96) [inline gateway/scim.py:253]: Why are we iterating over 1 element?

File: gateway/tests/test_scim.py [hunk 5]
```diff
@@ -2,6 +2,7 @@
 # All rights reserved.
 # See LICENSE file in this distribution.
 # SPDX-License-Identifier: Apache-2.0
+# pylint: disable=consider-using-with
```
Comment (rohitmohan96) [inline gateway/tests/test_scim.py:5]: Why is this required?

File: gateway/tests/test_scim.py [hunk 6]
```diff
@@ -14,6 +15,10 @@
 from testcontainers.postgres import PostgresContainer
 from app_factory import create_app
 from models import OnboardingAppKey
+# pylint: disable-next=unused-import
+from scim_fdo import FDOExtension
+# pylint: disable-next=unused-import
+from scim_ethermab import EtherMABExtension
```
Comment (rohitmohan96) [inline gateway/tests/test_scim.py:21]: This is similar to app.py. Why is this required and can we avoid it?

File: gateway/scim_extensions.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+"""
+
+This module implements SCIM extensions in a generic fashion.
+
+"""
+
+scim_ext_create = []
+scim_ext_read = []
+scim_ext_update = []
+scim_ext_delete = []
```
Comment (rohitmohan96) [inline gateway/scim_extensions.py:15]: I think this should be pre-populated with all the functions and the individual function should throw an error if `WANT_FDO` or similar is false.

File: gateway/tests/test_scim.py [hunk 8]
```diff
@@ -519,7 +527,7 @@ def test_delete_fdo_device(client: FlaskClient, api_key):
             "deviceDisplayName": "Generic FDO Device",
             "adminState": True,
             "urn:ietf:params:scim:schemas:extension:fido-device-onboard:2.0:Device": {
-                "fdoVoucher": "there-should-be-a-voucher-here"
+                "fdoOwnerVoucher": voucher
```
Comment (rohitmohan96) [inline gateway/tests/test_scim.py:530]: Can you extend the test to mock the FDO_OWNER_URI and intercept the request to verify that the request is what we expect? (We do something similar in the python sdk tests so you can take a look at that). 

I ask this because we are now sending a PEM and I would like to verify that the request is sent correctly with a test.

File: gateway/app.py (hunk 1 above)
Comment (elear) [reply gateway/app.py:34]: Because all database objects have to be declared so they can be instantiated in db.create_all.
This was driving me BONKERS until I figured it out.  And pytest won't spot it either.

File: gateway/models.py (hunk 3 above)
Comment (elear) [reply gateway/models.py:113]: I just haven't gotten there yet.

File: gateway/scim.py (hunk 4 above)
Comment (elear) [reply gateway/scim.py:253]: That's just leftovers before I do BLE, which requires its own PR because it's quite complicated.

File: gateway/tests/test_scim.py (hunk 5 above)
Comment (elear) [reply gateway/tests/test_scim.py:5]: See above

Comment (elear) [reply gateway/tests/test_scim.py:5]: Take it out and see what happens ;-)

File: gateway/tests/test_scim.py (hunk 6 above)
Comment (elear) [reply gateway/tests/test_scim.py:21]: Well, I can't avoid it but if you know of a way, great ;-)

File: gateway/scim_extensions.py (hunk 7 above)
Comment (elear) [reply gateway/scim_extensions.py:15]: If you do that, then you have to edit yet more files to add an extension.

File: gateway/tests/test_scim.py (hunk 8 above)
Comment (elear) [reply gateway/tests/test_scim.py:530]: I think that's possible, but I'd suggest opening an issue for it.

Comment (rohitmohan96) [reply gateway/tests/test_scim.py:530]: Opened #70

File: gateway/tests/test_scim.py (hunk 5 above)
Comment (rohitmohan96) [reply gateway/tests/test_scim.py:5]: Oh its because of 
```
voucher = open("tests/fdo-voucher.b64", encoding="utf-8").read()
```
Why didn't you use `with` instead of disabling it here?

File: gateway/tests/test_scim.py (hunk 8 above)
Comment (elear) [reply gateway/tests/test_scim.py:530]: I actually did a live test with the FDO OSS base.  I agree that building such a test is a good idea, but that's not for this ticket.

File: gateway/tests/test_scim.py (hunk 5 above)
Comment (elear) [reply gateway/tests/test_scim.py:5]: Because the error is spurious.  This does the job in one line instead of two, and it's a test, so there is no need to trap.

**Comment by elear** [comment]
@rohitmohan96 if you merge this one I'll start on the additional testing

**Review by rohitmohan96 — APPROVED** [review]
